File: Spring/Chapter4/java_demo/demo/src/main/java/com/aspectj_xml/XMLAspect.java
package com.aspectj_xml;

import org.aopalliance.intercept.Joinpoint;
import org.aspectj.lang.ProceedingJoinPoint;
import org.aspectj.lang.annotation.Before;

public class XMLAspect {
    public void before()
    {
        System.out.println("前置通知");
    }

    public void afterReturning()
    {
        System.out.println("后置返回通知");
    }

    public Object around(ProceedingJoinPoint proceedingJoinPoint) throws Throwable
    {
        System.out.println("环绕通知开始");
        Object object = proceedingJoinPoint.proceed();
        System.out.println("环绕通知结束");
        return object;
    }

    public void expect(Throwable e)
    {
        System.out.println("异常通知");
    }

    public void after()
    {
        System.out.println("后置最终通知");
    }
}

File: Spring/Chapter4/kotlin_demo/demo/src/main/kotlin/aspectj_xml/XMLAspect.kt
package aspectj_xml

import org.aspectj.lang.ProceedingJoinPoint
import org.springframework.aop.aspectj.MethodInvocationProceedingJoinPoint

class XMLAspect {
    fun before() = println("前置通知")

    fun afterReturning() = println("后置返回通知")

    fun around(proceedingJoinPoint: ProceedingJoinPoint):Any?
    {
        println("环绕通知开始")
        val obj = proceedingJoinPoint.proceed()
        println("环绕通知结束")
        return obj
    }

    fun expect(e:Throwable)
    {
        println("异常通知")
    }

    fun after()
    {
        println("后置返回通知")
    }
}